### PR TITLE
Add stream unzip

### PR DESCRIPTION
**Pull request by k-nasa**
ref: #129

### CODE REVIEW

File: src/stream/stream/unzip.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            stream,
+            res: (FromA::default(), FromB::default()),
+        }
+    }
+}
+
+impl<S, A, B, FromA, FromB> Future for UnzipFuture<S, FromA, FromB>
+where
+    S: Stream<Item = (A, B)>,
+    FromA: Default + Extend<A> + Copy,
```
Comment (yjhmelody) [inline src/stream/stream/unzip.rs]: Remove Copy
```suggestion
    FromA: Default + Extend<A>,
```

File: src/stream/stream/unzip.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    FromB: Default + Extend<B> + Copy,
```
Comment (yjhmelody) [inline src/stream/stream/unzip.rs]:
```suggestion
    FromB: Default + Extend<B>,
```

File: src/stream/stream/unzip.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+impl<S: Stream, FromA, FromB> UnzipFuture<S, FromA, FromB>
+where
+    FromA: Default,
+    FromB: Default,
+{
+    pub(super) fn new(stream: S) -> Self {
+        UnzipFuture {
+            stream,
+            res: (FromA::default(), FromB::default()),
```
Comment (yjhmelody) [inline src/stream/stream/unzip.rs]:
```suggestion
            res: Some((FromA::default(), FromB::default())),
```

File: src/stream/stream/unzip.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        let mut this = self.project();
+        let next = futures_core::ready!(this.stream.as_mut().poll_next(cx));
+
+        match next {
+            Some((a, b)) => {
+                this.res.0.extend(Some(a));
+                this.res.1.extend(Some(b));
+                Poll::Pending
+            }
+            None => Poll::Ready(*this.res),
```
Comment (yjhmelody) [inline src/stream/stream/unzip.rs]: use Option::take

Comment (yjhmelody) [reply src/stream/stream/unzip.rs]: You can reference #541 and std's unzip

Comment (k-nasa) [reply src/stream/stream/unzip.rs]: Good code! It became very helpful!

File: src/stream/stream/unzip.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        loop {
+            let next = futures_core::ready!(this.stream.as_mut().poll_next(cx));
+
+            match next {
+                Some((a, b)) => {
+                    let mut res = this.res.take().unwrap();
```
Comment (taiki-e) [inline src/stream/stream/unzip.rs]:
```suggestion
                    let res = this.res.as_mut().unwrap();
```

File: src/stream/stream/unzip.rs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                    res.0.extend(Some(a));
+                    res.1.extend(Some(b));
+
+                    *this.res = Some(res);
```
Comment (taiki-e) [inline src/stream/stream/unzip.rs]:
```suggestion
```

Comment (yjhmelody) [reply src/stream/stream/unzip.rs]: Well, I think my PR also need to change to `as_mut`

**Comment by k-nasa** [comment]
Oops! Unzip implementation PR has been submitted in the past. https://github.com/async-rs/async-std/pull/232

@tirr-c This PR seems to have stopped, what should I do?

**Comment by tirr-c** [comment]
@k-nasa I'll close mine -- returning two streams was pretty experimental, and now I think collecting with `Extend` is a better idea.

**Comment by k-nasa** [comment]
I like idea return two sterams. I was thinking same it.
"But is there a need to consume a stream and get a new stream?"
I thought, and made it as it is now.

File: src/stream/stream/mod.rs [hunk 7]
```diff
@@ -1717,6 +1719,44 @@ extension_trait! {
             Zip::new(self, other)
         }
 
+        #[doc = r#"
+            Converts an stream of pairs into a pair of containers.
+
+            unzip() consumes an entire stream of pairs, producing two collections: one from the left elements of the pairs, and one from the right elements.
```
Comment (taiki-e) [inline src/stream/stream/mod.rs]:
```suggestion
            `unzip()` consumes an entire stream of pairs, producing two collections: one from the left elements of the pairs, and one from the right elements.
```

Comment (k-nasa) [reply src/stream/stream/mod.rs]: Thanks

**Pull request merged**